### DISCUSSION TITLE
用PAG实现视频转场的几个疑问

### DISCUSSION THREAD

**Discussion by weizmann** [Tencent/libpag#1651] [Q&A] [answered]
你好，我在调研用PAG实现视频转场的效果，目前在调研实现以下[链接](https://zhuanlan.zhihu.com/p/70382363)中的特效

背景设定：
我在AE里面用了10s的前段视频+ 15s的后段视频作为占位符，然后导出PAG文件。
在PAGViewer里面看到的转场发生在**第10s**。

我的问题是：
1. 如果用户上传了一个20s的前段视频、20s的后段视频，PAG能否实现在前段视频播放完成后做转场？（即第20s左右开始转场，PAG文件中是在第10s左右开始转场）
>> 注意这里还是要按照用户视频的原速播放的，不能加快。。。

2. 如上面[zhihu链接](https://zhuanlan.zhihu.com/p/70382363)里面的『1. 白闪转场效果』：它是在前段后段视频中间增加了一个『纯色图层』。所以我感觉对于这种特效的话，应该把『纯色图层』单独作为一个PAG文件进行导出？然后用Composite进行组合嘛？

3. 还有一些，如『2. 翻页转场特效』：它是要求前后段视频有重叠的，所以应该在一个PAG文件中同时导出前段和后段视频？但是这样的话，又会存在我担心的问题，即：『如果填充视频和占位视频不一样长，是否能保证在填充视频结束后，播放转场特效』

我理解PAG应当是一个模板工具，那么他就不应当受填充物的影响？所以应该是有解法的么？
我是服务器端做渲染，资料和demo相对于移动端比较有限。希望您能在回答中给一些具体的代码 or 类相关的提点。

感谢感谢再感谢，感激万分~~

**Comment by domchen** [comment, marked as answer]
你想要的是运行时组合能力，我们在王者荣耀的游戏战报里也是这么用的，用户打出来的视频时长以及里面的高光时刻点都是不确定的，设计师在做的时候就分开导出一个小的PAG文件，每个PAG只管一个效果，比如一个震屏，一个转场。运行时你可以直接用PAGFile.setDuration()+setTimeStretchMode（）来按照想要的方式拉伸一个PAG，最后创建一个空的PAGComposition，把他们摆放在一起就可以了。通过PAGLayer的setStartTime可以控制时间轴上的起始位置。具体可以参考这个demo：https://github.com/libpag/pag-ios/blob/master/sample/pag-ios/ViewController/RenderMultiplePAGFilesViewController.m

**Reply by EngeWang** [reply]
如果我制作两个 pag 文件，前一个是单独的视频，后一个是带 bmp 的遮罩视频及占位，进行转场。为什么往后叠加，会出现黑的情况，如第一个 pag 跟第二个 pag叠加无问题，到了第二个和第三个叠加时出现黑屏。是什么问题。未使用 sdk
